mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAOV2.java

```diff
@@ -0,0 +1,360 @@
+                .setTimestamp(INTERNAL_DATE, message.getInternalDate())
+                .setInt(BODY_START_OCTET, (int) (message.getFullContentOctets() - message.getBodyOctets()))
+                .setLong(FULL_CONTENT_OCTETS, message.getFullContentOctets())
+                .setLong(BODY_OCTECTS, message.getBodyOctets())
+                .setUUID(BODY_CONTENT, bodyContentId.orElse(null))
+                .setUUID(HEADER_CONTENT, headerContentId.orElse(null))
+                .setList(PROPERTIES, message.getProperties().stream()
+                        .map(x -> typesProvider.getDefinedUserType(PROPERTIES)
+                                .newValue()
+                                .setString(Properties.NAMESPACE, x.getNamespace())
+                                .setString(Properties.NAME, x.getLocalName())
+                                .setString(Properties.VALUE, x.getValue()))
+                        .collect(Collectors.toList()))
+                .setList(ATTACHMENTS, message.getAttachments().stream()
+                        .map(this::toUDT)
+                        .collect(Guavate.toImmutableList()));
+    }
+
+    private BoundStatement setTextualLineCount(BoundStatement boundStatement, Long textualLineCount) {
+        return Optional.ofNullable(textualLineCount)
+               .map(value -> boundStatement.setLong(TEXTUAL_LINE_COUNT, value))
+               .orElseGet(() -> boundStatement.setToNull(TEXTUAL_LINE_COUNT));
+    }
+
+    private UDTValue toUDT(MessageAttachment messageAttachment) {
+        return typesProvider.getDefinedUserType(ATTACHMENTS)
+            .newValue()
+            .setString(Attachments.ID, messageAttachment.getAttachmentId().getId())
+            .setString(Attachments.NAME, messageAttachment.getName().orNull())
+            .setString(Attachments.CID, messageAttachment.getCid().transform(Cid::getValue).orNull())
+            .setBool(Attachments.IS_INLINE, messageAttachment.isInline());
+    }
+
+    public CompletableFuture<Stream<Pair<MessageWithoutAttachment, Stream<MessageAttachmentRepresentation>>>> retrieveMessages(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
+        return CompletableFutureUtil.chainAll(
+            getLimitedIdStream(messageIds.stream().distinct(), limit)
+                .collect(JamesCollectors.chunker(CHUNK_SIZE_ON_READ)),
+            ids -> FluentFutureStream.of(
+                ids.stream()
+                    .map(id -> retrieveRow(id, fetchType)
+                        .thenApply((ResultSet resultSet) ->
+                            message(resultSet.one(), id, fetchType))))
+                .completableFuture())
+            .thenApply(stream -> stream.flatMap(Function.identity()));
+    }
+
+    private Stream<ComposedMessageIdWithMetaData> getLimitedIdStream(Stream<ComposedMessageIdWithMetaData> messageIds, Optional<Integer> limit) {
+        return limit
+            .filter(value -> value > 0)
```

Comment:
Negative limit means unlimited. It's ugly but it's an heritage from the mailbox api